pootle/apps/pootle_store/diff.py

```diff
@@ -0,0 +1,235 @@
+        new_unitid_list = self.new_unit_list
+        for (tag, i1, i2, j1, j2) in self.opcodes:
+            if tag == 'insert':
+                update_index_delta = 0
+                previous_index = 0
+                if i1 > 0:
+                    previous = self.db_units[self.active_units[i1 - 1]]
+                    previous_index = previous['index']
+                next_index = previous_index + 1
+                if i1 < len(self.active_units):
+                    next = self.db_units[self.active_units[i1]]
+                    next_index = next['index']
+                    delta = j2 - j1
+                    update_index_delta = (
+                        delta - next_index + previous_index + 1)
+
+                inserts.append((previous_index,
+                                new_unitid_list[j1:j2],
+                                next_index,
+                                update_index_delta))
+
+            elif tag == 'replace':
+                i1_index = self.db_units[self.active_units[i1 - 1]]['index']
+                i2_index = self.db_units[self.active_units[i2 - 1]]['index']
+                update_index_delta = j2 - j1 - i2_index + i1_index
+                inserts.append((i1_index,
+                                new_unitid_list[j1:j2],
+                                i2_index,
+                                update_index_delta))
+        return inserts
+
+    @cached_property
+    def new_unit_list(self):
+        # If file_revision is gte than the db_revision then new unit list
+        # will be exactly what is in the file
+        if self.file_revision >= self.db_revision:
+            return self.file_unit_ids
+
+        # These units are kept as they have been updated since file_revision
+        # but do not appear in the file
+        new_units = [u for u in self.updated_db_units
+                     if u not in self.file_unit_ids]
+
+        # These unit are either present in both or only in the file so are
+        # kept in the file order
+        new_units += [u for u in self.file_unit_ids
+                      if u not in self.obsoleted_db_units]
+
+        return new_units
+
+    @cached_property
+    def obsoleted_db_units(self):
+        return [unitid for unitid, unit in self.db_units.items()
+                if unit['state'] == OBSOLETE
+                and unit["revision"] > self.file_revision]
+
+    @cached_property
+    def opcodes(self):
+        sm = difflib.SequenceMatcher(None, self.active_units,
+                                     self.new_unit_list)
+        return sm.get_opcodes()
+
+    @cached_property
+    def updated_db_units(self):
+        return [unitid for unitid, unit in self.db_units.items()
+                if unit['revision'] > self.file_revision
+                and not unit["state"] == OBSOLETE]
+
+    def get_updated_dbids(self):
+        """Returns a set of unit DB ids to be updated.
+        """
+        update_dbids = set()
+
+        def _has_changed(uid):
```

Comment:
@phlax, I found the reason why we have a speedup. Number of updated units in [this PR (pootle_store/models.py#L1614)] (https://github.com/phlax/pootle/blob/aecd9338daa6e3f987bccfdfe4230aaa3742a216/pootle/apps/pootle_store/models.py#L1614) differs from [current HEAD (pootle_store/models.py#L1689)](https://github.com/translate/pootle/blob/f7f17237fa8f10ebfaf8366a76843c1ee2826883/pootle/apps/pootle_store/models.py#L1689)
You filter unit list here and return `changed` units only. And you don't take into account `state`, `developer_comment`, `translator_comment`, `locations`, `context` unit fields (see [unit.update](https://github.com/phlax/pootle/blob/aecd9338daa6e3f987bccfdfe4230aaa3742a216/pootle/apps/pootle_store/models.py#L695))

And yeah, it's very strange that don't have tests for this.